### PR TITLE
Alloc domain patch concretization

### PR DESCRIPTION
**Pull request by jjsjann123**
DynamicTransformConcretizer should preserve allocation domain

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

File: python_tests/test_python_frontend.py [hunk 1]
```diff
@@ -2584,6 +2584,29 @@ def dynamic_reshape(fd: FusionDefinition) -> None:
         self.assertEqual(y.shape, torch.Size([3, 2, 2]))
         self.assertEqual(x.flatten(), y.flatten())
 
+    def test_allocation_domain_concretization(self):
+        inputs = [
+            torch.randn((0,), dtype=torch.float64, device="cuda:0").as_strided(
+                (1, 0, 1, 1), (0, 1, 1, 1)
```
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:2590]: Why does the input need to be empty here?

File: python_tests/test_python_frontend.py [hunk 2]
```diff
@@ -2584,6 +2584,29 @@ def dynamic_reshape(fd: FusionDefinition) -> None:
...
+                contiguity=[True, None, None, None],
+                dtype=DataType.Double,
+                is_cpu=False,
+                stride_order=[0, 3, 2, 1],
+            )
+            S1 = fd.define_scalar(2.0, dtype=DataType.Double)
+            T2 = fd.ops.mul(T1, S1)
+            fd.add_output(T2)
+
+        nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
```
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:2606]: Does this check strides?

**Review by jacobhinkle — APPROVED** [review]
Looks great. Thanks for the update!

File: python_tests/test_python_frontend.py (hunk 2 above)
Comment (jjsjann123) [reply python_tests/test_python_frontend.py:2606]: I don't think the stride matters here. i.e. we are not specifying stride on output, so codegen is free to do what it sees fits.

File: python_tests/test_python_frontend.py (hunk 1 above)
Comment (jjsjann123) [reply python_tests/test_python_frontend.py:2590]: I thought I needed this in order to trigger concretization, i.e. `concretizeEmptyExtents` in order to enter the code path.

**Comment by jjsjann123** [comment]
looks like we accidentally added bfloat16 on python test, which is now failing on v100. I'll patch that.

**Comment by jjsjann123** [comment]
Patched in #1365.

This PR should be good by itself. I'm merging with broken CI.

File: python_tests/test_python_frontend.py (hunk 1 above)
Comment (wujingyue) [reply python_tests/test_python_frontend.py:2590]: I had the same question :) I think it's worth at least a code comment. It'd be even better if we can verify concretization is indeed triggered. This way, **if** later we no longer concretize empty extents, we'll get notified and change the test to trigger concretization in a different way.

File: csrc/dynamic_transform.cpp [hunk 3]
```diff
@@ -753,6 +753,9 @@ void DynamicTransformConcretizer::mutate(TensorDomain* td) {
       ? updateIdVec(td->maybeRFactor())
       : std::vector<IterDomain*>();
   std::vector<IterDomain*> domain = updateIdVec(td->leaf());
```
Comment (wujingyue) [inline csrc/dynamic_transform.cpp:755]:
```suggestion
  std::vector<IterDomain*> leaf_domain = updateIdVec(td->leaf());
```

File: csrc/dynamic_transform.cpp [hunk 4]
```diff
@@ -753,6 +753,9 @@ void DynamicTransformConcretizer::mutate(TensorDomain* td) {
...
+  std::vector<IterDomain*> alloc_dom = td->hasAllocation()
+      ? updateIdVec(td->maybeAllocation())
```
Comment (wujingyue) [inline csrc/dynamic_transform.cpp:757]:
```suggestion
      ? updateIdVec(td->allocation())
```

File: csrc/dynamic_transform.cpp [hunk 5]
```diff
@@ -753,6 +753,9 @@ void DynamicTransformConcretizer::mutate(TensorDomain* td) {
       ? updateIdVec(td->maybeRFactor())
```
Comment (wujingyue) [inline csrc/dynamic_transform.cpp:753]:
```suggestion
      ? updateIdVec(td->rfactor())
```

File: csrc/dynamic_transform.cpp [hunk 6]
```diff
@@ -761,8 +764,16 @@ void DynamicTransformConcretizer::mutate(TensorDomain* td) {
   // Update the contiguity vector. Drop the contig val if mutated to broadcast
   auto contig = td->contiguity();
 
-  for (const auto i : c10::irange(td->maybeRFactor().size())) {
-    auto original_id = td->maybeRFactor().at(i);
+  const auto& new_alloc = td->hasAllocation() ? alloc_dom
```
Comment (wujingyue) [inline csrc/dynamic_transform.cpp:767]:
```suggestion
  const auto& new_maybe_alloc = td->hasAllocation() ? alloc_dom
```

to distinguis it from `alloc_dom`, the real new allocation domain.

**Review by wujingyue — APPROVED** [review]

File: python_tests/test_python_frontend.py (hunk 1 above)
Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:2590]: > I thought I needed this in order to trigger concretization, i.e. `concretizeEmptyExtents` in order to enter the code path.

I see. Yes that works. From C++ we could use a dynamic reshape but this way exercises the python alloc domain stuff too.

**Pull request merged**